File: rca/rca-engine/src/rules/vendor_anchors.py
"""Vendor-aware anchor helpers for reducing cross-vendor rule noise."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from src.utils.dicts import get_nested


_RULE_TAG_TO_VENDOR: dict[str, str] = {
    "cisco": "cisco",
    "ios": "cisco",
    "nxos": "cisco",
    "juniper": "juniper",
    "junos": "juniper",
    "arista": "arista",
    "eos": "arista",
    "aruba": "aruba",
    "hpe": "aruba",
    "hewlett_packard_enterprise": "aruba",
    "aos": "aruba",
    "aoscx": "aruba",
    "aos_switch": "aruba",
    "procurve": "aruba",
    "huawei": "huawei",
    "vrp": "huawei",
    "cloudengine": "huawei",
    "netengine": "huawei",
    "quidway": "huawei",
    "dell": "dell",
    "os10": "dell",
    "powerswitch": "dell",
    "smartfabric": "dell",
    "dellos": "dell",
    "fortinet": "fortinet",
    "fortios": "fortinet",
    "fortigate": "fortinet",
    "checkpoint": "checkpoint",
    "paloalto": "paloalto",
    "panos": "paloalto",
    "f5": "f5",
    "bigip": "f5",
    "mikrotik": "mikrotik",
    "routeros": "mikrotik",
}

_VENDOR_TOKENS: dict[str, tuple[str, ...]] = {
    "cisco": ("cisco", "nx-os", "nxos", "ios", "cat9k", "c9300", "isr"),
    "juniper": ("juniper", "junos", "srx", "qfx", "mx", "ex"),
    "arista": ("arista", "eos"),
    "aruba": (
        "aruba",
        "aruba networks",
        "aruba networking",
        "hpe",
        "hpe aruba",
        "hpe aruba networking",
        "hewlett packard enterprise",
        "arubaos",
        "aos-cx",
        "aoscx",
        "aos-switch",
        "aos switch",
        "procurve",
        "aruba central",
    ),
    "huawei": (
        "huawei",
        "huawei technologies",
        "vrp",
        "cloudengine",
        "ce12800",
        "ce6800",
        "ce5800",
        "netengine",
        "ne40e",
        "quidway",
    ),
    "dell": (
        "dell",
        "dell emc",
        "powerswitch",
        "smartfabric",
        "smartfabric os10",
        "os10",
        "dellos",
        "n-series",
        "s-series",
        "z-series",
    ),
    "fortinet": ("fortinet", "fortigate", "fortios", "fgt"),
    "checkpoint": ("checkpoint", "check point", "cplogexporter", "gaia"),
    "paloalto": ("paloalto", "palo alto", "pan-os", "panos", "pa-vm"),
    "f5": ("f5", "big-ip", "bigip", "ltm", "mcpd", "tmm"),
    "mikrotik": ("mikrotik", "routeros", "ros", "mtk"),
}

_STRICT_HINT_FIELDS: tuple[str, ...] = (
    "observer.vendor",
    "observer.product",
    "observer.type",
    "observer.name",
    "observer.hostname",
    "event.module",
    "event.dataset",
    "data_stream.dataset",
    "device.vendor",
    "device.type",
    "device.model",
    "host.name",
    "host.hostname",
    "agent.name",
    "agent.type",
)

_SOFT_HINT_FIELDS: tuple[str, ...] = (
    "message",
    "msg",
    "event.original",
)


def infer_rule_vendor(tags: list[str]) -> str | None:
    """Return canonical vendor key inferred from rule tags."""
    for raw_tag in tags:
        tag = str(raw_tag).strip().lower()
        if not tag:
            continue
        vendor = _RULE_TAG_TO_VENDOR.get(tag)
        if vendor is not None:
            return vendor
    return None


@dataclass(slots=True)
class VendorAnchorSnapshot:
    """Normalized vendor-hint cache for one source event."""

    strict_values: tuple[str, ...]
    soft_values: tuple[str, ...]

    @property
    def has_strict_hints(self) -> bool:
        """Return True when event includes explicit vendor metadata."""
        return bool(self.strict_values)

    @classmethod
    def from_event(cls, event: dict[str, Any]) -> "VendorAnchorSnapshot":
        """Build snapshot by extracting and normalizing vendor hint fields."""
        strict_values = tuple(_collect_field_values(event, _STRICT_HINT_FIELDS))
        soft_values = tuple(_collect_field_values(event, _SOFT_HINT_FIELDS))
        return cls(strict_values=strict_values, soft_values=soft_values)

    def matches_vendor(self, vendor: str) -> bool:
        """Return True when event hints align with the provided vendor key."""
        tokens = _VENDOR_TOKENS.get(vendor)
        if not tokens:
            return False
        # When strict metadata exists (observer/device/agent fields), treat it as source
        # of truth and ignore soft text hints to avoid cross-vendor false positives.
        values = self.strict_values if self.strict_values else self.soft_values
        for value in values:
            for token in tokens:
                if token in value:
                    return True
        return False


def _collect_field_values(event: dict[str, Any], fields: tuple[str, ...]) -> list[str]:
    """Collect lowercase string values for selected event fields."""
    values: list[str] = []
    for field in fields:
        raw = get_nested(event, field)
        if raw is None and "." not in field:
            raw = event.get(field)
        values.extend(_normalize_value(raw))
    return values


def _normalize_value(raw: Any) -> list[str]:
    """Convert scalar/list values to lowercase strings for token matching."""
    if raw is None:
        return []
    if isinstance(raw, (str, int, float, bool)):
        text = str(raw).strip().lower()
        return [text] if text else []
    if isinstance(raw, list):
        out: list[str] = []
        for item in raw:
            out.extend(_normalize_value(item))
        return out
    return []

File: rca/log_signalizing/signalizing_go/internal/rca/rules/vendor_anchors.go
package rules

import (
	"fmt"
	"strings"

	"rca/internal/rca/util"
)

var ruleTagToVendor = map[string]string{
	"cisco":                      "cisco",
	"ios":                        "cisco",
	"nxos":                       "cisco",
	"juniper":                    "juniper",
	"junos":                      "juniper",
	"arista":                     "arista",
	"eos":                        "arista",
	"aruba":                      "aruba",
	"hpe":                        "aruba",
	"hewlett_packard_enterprise": "aruba",
	"aos":                        "aruba",
	"aoscx":                      "aruba",
	"aos_switch":                 "aruba",
	"procurve":                   "aruba",
	"huawei":                     "huawei",
	"vrp":                        "huawei",
	"cloudengine":                "huawei",
	"netengine":                  "huawei",
	"quidway":                    "huawei",
	"dell":                       "dell",
	"os10":                       "dell",
	"powerswitch":                "dell",
	"smartfabric":                "dell",
	"dellos":                     "dell",
	"fortinet":                   "fortinet",
	"fortios":                    "fortinet",
	"fortigate":                  "fortinet",
	"checkpoint":                 "checkpoint",
	"paloalto":                   "paloalto",
	"panos":                      "paloalto",
	"f5":                         "f5",
	"bigip":                      "f5",
	"mikrotik":                   "mikrotik",
	"routeros":                   "mikrotik",
}

var vendorTokens = map[string][]string{
	"cisco":     {"cisco", "nx-os", "nxos", "ios", "cat9k", "c9300", "isr"},
	"juniper":   {"juniper", "junos", "srx", "qfx", "mx", "ex"},
	"arista":    {"arista", "eos"},
	"aruba":     {"aruba", "aruba networks", "aruba networking", "hpe", "hpe aruba", "hpe aruba networking", "hewlett packard enterprise", "arubaos", "aos-cx", "aoscx", "aos-switch", "aos switch", "procurve", "aruba central"},
	"huawei":    {"huawei", "huawei technologies", "vrp", "cloudengine", "ce12800", "ce6800", "ce5800", "netengine", "ne40e", "quidway"},
	"dell":      {"dell", "dell emc", "powerswitch", "smartfabric", "smartfabric os10", "os10", "dellos", "n-series", "s-series", "z-series"},
	"fortinet":  {"fortinet", "fortigate", "fortios", "fgt"},
	"checkpoint": {"checkpoint", "check point", "cplogexporter", "gaia"},
	"paloalto":  {"paloalto", "palo alto", "pan-os", "panos", "pa-vm"},
	"f5":        {"f5", "big-ip", "bigip", "ltm", "mcpd", "tmm"},
	"mikrotik":  {"mikrotik", "routeros", "ros", "mtk"},
}

var strictHintFields = []string{
	"observer.vendor",
	"observer.product",
	"observer.type",
	"observer.name",
	"observer.hostname",
	"event.module",
	"event.dataset",
	"data_stream.dataset",
	"device.vendor",
	"device.type",
	"device.model",
	"host.name",
	"host.hostname",
	"agent.name",
	"agent.type",
}

var softHintFields = []string{
	"message",
	"msg",
	"event.original",
}

// VendorAnchorSnapshot is the normalized vendor hint cache for one event.
type VendorAnchorSnapshot struct {
	StrictValues []string
	SoftValues   []string
}

// HasStrictHints returns true when event includes explicit vendor metadata.
func (s VendorAnchorSnapshot) HasStrictHints() bool {
	return len(s.StrictValues) > 0
}

// MatchesVendor returns true when the event hints align with the vendor.
func (s VendorAnchorSnapshot) MatchesVendor(vendor string) bool {
	tokens := vendorTokens[vendor]
	if len(tokens) == 0 {
		return false
	}

	values := s.SoftValues
	if len(s.StrictValues) > 0 {
		values = s.StrictValues
	}

	for _, value := range values {
		for _, token := range tokens {
			if strings.Contains(value, token) {
				return true
			}
		}
	}
	return false
}

// InferRuleVendor returns the canonical vendor inferred from rule tags.
func InferRuleVendor(tags []string) string {
	for _, rawTag := range tags {
		tag := strings.TrimSpace(strings.ToLower(rawTag))
		if tag == "" {
			continue
		}
		if vendor, ok := ruleTagToVendor[tag]; ok {
			return vendor
		}
	}
	return ""
}

// VendorAnchorSnapshotFromEvent builds a snapshot by normalizing vendor-hint fields.
func VendorAnchorSnapshotFromEvent(event map[string]any) VendorAnchorSnapshot {
	return VendorAnchorSnapshot{
		StrictValues: collectFieldValues(event, strictHintFields),
		SoftValues:   collectFieldValues(event, softHintFields),
	}
}

func collectFieldValues(event map[string]any, fields []string) []string {
	values := make([]string, 0)
	for _, field := range fields {
		raw := util.GetNested(event, field)
		if raw == nil && !strings.Contains(field, ".") {
			raw = event[field]
		}
		values = append(values, normalizeVendorValue(raw)...)
	}
	return values
}

func normalizeVendorValue(raw any) []string {
	switch typed := raw.(type) {
	case nil:
		return nil
	case string:
		text := strings.TrimSpace(strings.ToLower(typed))
		if text == "" {
			return nil
		}
		return []string{text}
	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64, bool:
		text := strings.TrimSpace(strings.ToLower(fmt.Sprint(typed)))
		if text == "" {
			return nil
		}
		return []string{text}
	case []any:
		values := make([]string, 0)
		for _, item := range typed {
			values = append(values, normalizeVendorValue(item)...)
		}
		return values
	case []string:
		values := make([]string, 0, len(typed))
		for _, item := range typed {
			values = append(values, normalizeVendorValue(item)...)
		}
		return values
	default:
		return nil
	}
}
